File: python/p0001_two_sum/solutions/recursion_iteration.py
# Time complexity: O(n^2)
# Space complexity: O(n)

from typing import List


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        
        def find_indices(i: int) -> List[int]:
            
            # Base case
            if i >= len(nums):
                return [-1]
            
            # Recursive case
            for j in range(i + 1, len(nums)):
                if nums[i] + nums[j] == target:
                    return [i, j]
            
            return find_indices(i + 1)
        
        return find_indices(0)

File: rust/p0001_two_sum/src/recursion_iteration.rs
// Time complexity: O(n^2)
// Space complexity: O(n)

pub struct Solution;

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        fn find_indices(nums: &[i32], target: i32, i: usize) -> Vec<i32> {
            // Base case
            if i >= nums.len() {
                return vec![-1];
            }

            // Recursive case
            for j in i + 1..nums.len() {
                if nums[i] + nums[j] == target {
                    return vec![i as i32, j as i32];
                }
            }

            find_indices(&nums, target, i + 1)
        }

        find_indices(&nums, target, 0)
    }
}
